### PR TITLE
feat: generalize SplitPredicate to functions, and more

### PR DESCRIPTION
**Pull request by TWal**
In the aim to use the SplitPredicate module for KDF invariants, that are functions outputting `usage` or `label`, I first generalized the module to support functions (noting that predicates are functions to `prop`).

I also generalized the tag encoding mechanism so that local functions are associated with a set of tags. I hoped it would be useful to split the invariants of KDF.Extract, but now I'm not convinced it would be useful toward that goal. Anyway, I feel this is still an improvement to the `encode_tag` mechanism that felt quite ad-hoc (because it was), this mechanism is strictly more generic, by using the following translation:

```fstar
let tag_set_t = old.tag_t
let tag_t = old.encoded_tag_t
let tag_belong_to tag tag_set =
  tag = old.encode_tag tag_set
let cant_belong_to_disjoint_sets tag tag_set_1 tag_set_2 =
  FStar.Classical.move_requires_3 old.encode_tag_inj tag_set_1 tag_set_2
```

meaning that the `encode_tag` mechanism can only be used to represent single-element tag sets.

In the process, I also greatly simplified the proofs of the SplitFunction module!

### CODE REVIEW

**Comment by TWal** [comment]
@cwaldm do you have any thoughts on this, since you have been looking closely at this file? I would like to define functions in a modular way soon, hence merging this PR would be useful.

File: src/lib/utils/DY.Lib.SplitFunction.fst [hunk 1]
```diff
@@ -0,0 +1,283 @@
+module DY.Lib.SplitFunction
+
+open FStar.List.Tot
+open Comparse //for_allP
```
Comment (cwaldm) [inline src/lib/utils/DY.Lib.SplitFunction.fst:4]: can we move `for_allP` and `for_allP_eq` to somewhere else, so that we do not have to call comparse here?
after all this module is completely unrelated to comparse.
(not a major problem, we can also keep it like this, but then we should add `for_allP_eq` to the comment as well)

File: src/lib/utils/DY.Lib.SplitFunction.fst [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+val has_local_fun: func:split_function_input_values -> func.global_fun -> (func.tag_set_t & func.local_fun) -> prop
+let has_local_fun func gfun (tag_set, lfun) =
+  forall tagged_data.
+    match func.decode_tagged_data tagged_data with
+    | Some (tag, raw_data) ->
+      tag `func.tag_belong_to` tag_set ==> (func.apply_global_fun gfun tagged_data == func.apply_local_fun lfun raw_data)
+    | None -> True
+
+val find_local_fun: func:split_function_input_values -> list (func.tag_set_t & func.local_fun) -> func.tag_t -> option func.local_fun
```
Comment (cwaldm) [inline src/lib/utils/DY.Lib.SplitFunction.fst:125]: maybe add a comment on what this function does.
something along the lines of "checks if a given tag appears in a collection of tagged local functions and if so, returns the associated local function".
(is it important to mention that it returns the _first_ local function where the tag appears? in our use-cases it is probably unique since we have the disjointness property?)

File: src/lib/event/DY.Lib.Event.Typed.fst [hunk 3]
```diff
@@ -41,35 +41,40 @@ let mk_event_instance #a #format tag = {
...
   tag_t = string;
-  encoded_tag_t = string;
+  is_disjoint = default_disjoint;
+  tag_belong_to = (fun dtag tag -> dtag = tag);
+  cant_belong_to_disjoint_sets = (fun dtag tag1 tag2 -> ());
+
   raw_data_t = trace & principal & bytes;
+  output_t = prop;
+
+  default_global_fun = (fun tr prin tag content -> False);
```
Comment (cwaldm) [inline src/lib/event/DY.Lib.Event.Typed.fst:56]: i think this should come after `global_fun` to match the order in the definition of the `split_function_input_values`, and also to know the type of this function before defining it.

File: src/lib/state/DY.Lib.State.Tagged.fst [hunk 4]
```diff
@@ -40,45 +40,50 @@ type local_bytes_state_predicate {|crypto_invariants|} = {
...
   tag_t = string;
-  encoded_tag_t = string;
+  is_disjoint = default_disjoint;
+  tag_belong_to = (fun dtag tag -> dtag = tag);
+  cant_belong_to_disjoint_sets = (fun dtag tag1 tag2 -> ());
+
   raw_data_t = trace & principal & state_id & bytes;
+  output_t = prop;
+
+  default_global_fun = (fun tr prin sess_id sess_content -> False);
```
Comment (cwaldm) [inline src/lib/state/DY.Lib.State.Tagged.fst:55]: this should move after `global_fun`. (same as for `DY.Lib.Event.Typed`)

File: src/lib/state/DY.Lib.State.Tagged.fst [hunk 5]
```diff
@@ -87,27 +92,24 @@ val mk_global_local_bytes_state_predicate_correct: invs:protocol_invariants -> l
...
-    mk_global_pred_eq split_local_bytes_state_predicate_func lpreds (tr2, prin, sess_id, full_content);
-    assert(split_local_bytes_state_predicate_func.apply_local_pred lpred (tr2, prin, sess_id, content))
-  )
+  mk_global_fun_eq split_local_bytes_state_predicate_func lpreds (tr1, prin, sess_id, full_content);
+  mk_global_fun_eq split_local_bytes_state_predicate_func lpreds (tr2, prin, sess_id, full_content);
+  introduce forall lpred content. split_local_bytes_state_predicate_func.apply_local_fun lpred (tr1, prin, sess_id, content) ==> split_local_bytes_state_predicate_func.apply_local_fun lpred (tr2, prin, sess_id, content) with (
+    introduce _ ==> _ with _. lpred.pred_later tr1 tr2 prin sess_id content
+  );
+  match split_local_bytes_state_predicate_func.decode_tagged_data (tr1, prin, sess_id, full_content) with
+  | Some (_, (_, _, _, _)) -> ()
```
Comment (cwaldm) [inline src/lib/state/DY.Lib.State.Tagged.fst:111]: i think this pattern can be simplified to `Some (_,_)` .
(but interestingly not to just `Some _` ?!)

File: src/lib/utils/DY.Lib.SplitFunction.fst [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    has_local_fun func gfun (tag_set, lfun)
+  )
+  (ensures func.apply_global_fun gfun tagged_data == func.apply_local_fun lfun raw_data)
+let local_eq_global_lemma func gfun tag_set lfun tagged_data tag raw_data = ()
+
+/// In the common case where tag disjointness is unequality,
+/// `all_disjoint` is equivalent to the `no_repeats_p` predicate
+/// from F*'s standard library.
+
+val default_disjoint:
```
Comment (cwaldm) [inline src/lib/utils/DY.Lib.SplitFunction.fst:267]: i'd prefer if this name includes "inequality" or "unequal", so that when we use this function (for example in Events or State.Tagged) one can immediately see what it means.
I can think of several options "default_disjoint_inequality", "disjoint_inequality", "inequality_disjoint", ...
Or maybe even just "unequal"? Then it would read nicely in, say, Event: `is_disjoint = unequal;`

File: src/lib/utils/DY.Lib.SplitFunction.fst [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+
+  // Input type for the local functions
+  raw_data_t: Type;
+
+  // Output type for the local and global functions
+  output_t: Type;
+
+  // We can decode the global function input
+  // into an encoded tag and a local function input
```
[the remark this answers is not in the document]
Comment (TWal) [reply src/lib/utils/DY.Lib.SplitFunction.fst]: Indeed, I removed that word

File: src/lib/event/DY.Lib.Event.Typed.fst (hunk 3 above)
Comment (TWal) [reply src/lib/event/DY.Lib.Event.Typed.fst]: Good catch, I changed that!

File: src/lib/state/DY.Lib.State.Tagged.fst (hunk 4 above)
Comment (TWal) [reply src/lib/state/DY.Lib.State.Tagged.fst]: Good catch, I changed that!

File: src/lib/utils/DY.Lib.SplitFunction.fst (hunk 6 above)
Comment (TWal) [reply src/lib/utils/DY.Lib.SplitFunction.fst]: I went for `unequal`, this is a better name indeed

File: src/lib/utils/DY.Lib.SplitFunction.fst (hunk 1 above)
Comment (TWal) [reply src/lib/utils/DY.Lib.SplitFunction.fst]: I agree this is inconvenient. For now I went for the better comment.
I see three possibilities, from best to worst (in my opinion):
- either we try to move `for_allP` in F*'s standard library, which is inconvenient because this requires to do a pull-request
- either we keep using Comparse's `for_allP` because it's already there, which is inconvenient in modules like this one that shouldn't depend on Comparse
- either we re-implement `for_allP` in DY* in some module `DY.Utils` or something, which is inconvenient because this will cause weird shadowing issues with Comparse's `for_allP`

File: src/lib/utils/DY.Lib.SplitFunction.fst (hunk 2 above)
Comment (TWal) [reply src/lib/utils/DY.Lib.SplitFunction.fst]: Done, I hope it is clear enough!

File: src/lib/utils/DY.Lib.SplitFunction.fst [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+val has_local_fun: func:split_function_input_values -> func.global_fun -> (func.tag_set_t & func.local_fun) -> prop
+let has_local_fun func gfun (tag_set, lfun) =
+  forall tagged_data.
+    match func.decode_tagged_data tagged_data with
+    | Some (tag, raw_data) ->
+      tag `func.tag_belong_to` tag_set ==> (func.apply_global_fun gfun tagged_data == func.apply_local_fun lfun raw_data)
+    | None -> True
+
+val find_local_fun: func:split_function_input_values -> list (func.tag_set_t & func.local_fun) -> func.tag_t -> option func.local_fun
+let rec find_local_fun func l tag =
```
[the remark this answers is not in the document]
Comment (TWal) [reply src/lib/utils/DY.Lib.SplitFunction.fst]: Correct, I went for the more explicit name `tagged_local_funs`

File: src/lib/state/DY.Lib.State.Tagged.fst (hunk 5 above)
Comment (TWal) [reply src/lib/state/DY.Lib.State.Tagged.fst:111]: Indeed, I think we would need to increase `ifuel` to have `Some _` (this is probably explained in FStarLang/FStar#3076).
I would prefer keeping the general structure, as some kind of documentation on the type of thing being matched. I think I didn't put the names here because they would mostly be the same as variables in the context (e.g. `prin` or `sess_id`), but we could write them to make things more explicit.

File: src/lib/utils/DY.Lib.SplitFunction.fst [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+  (requires List.Tot.memP x l)
+  (ensures List.Tot.memP (f x) (List.Tot.map f l))
+let rec memP_map #a #b f l x =
+  match l with
+  | [] -> ()
+  | h::t ->
+    introduce x =!= h ==> List.Tot.memP (f x) (List.Tot.map f t)
+    with _. memP_map f t x
+
+val all_disjoint : #a:Type -> (a -> a -> prop) -> list a -> prop
```
Comment (cwaldm) [inline src/lib/utils/DY.Lib.SplitFunction.fst:170]: i think this function should have a more general name since it already has the general type.
maybe something like "for_allP_pairwise" ? (then it reads nicely when calling `for_allP_pairwise is_disjoint ...`)

File: src/lib/utils/DY.Lib.SplitFunction.fst [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+  (requires
+    func.decode_tagged_data tagged_data == Some (tag, raw_data) /\
+    tag `func.tag_belong_to` tag_set /\
+    has_local_fun func global_fun (tag_set, local_fun)
+  )
+  (ensures func.apply_global_fun global_fun tagged_data == func.apply_local_fun local_fun raw_data)
+let local_eq_global_lemma func global_fun tag_set tagged_local_fun tagged_data tag raw_data = ()
+
+/// In the common case where tag disjointness is unequality,
+/// `all_disjoint` is equivalent to the `no_repeats_p` predicate
```
Comment (cwaldm) [inline src/lib/utils/DY.Lib.SplitFunction.fst:268]: the comment says "equivalent" but the lemma only shows "no_repeats ==> all unequal".
can we show equivalence?

File: src/lib/utils/DY.Lib.SplitFunction.fst [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+/// from F*'s standard library.
+
+val unequal:
+  #a:Type ->
+  a -> a -> prop
+let unequal #a x y = x =!= y
+
+val no_repeats_p_implies_all_disjoint:
```
Comment (cwaldm) [inline src/lib/utils/DY.Lib.SplitFunction.fst:276]: i think this name should contain unequal, since it is only true for "all_disjoint unequal"

File: src/lib/utils/DY.Lib.SplitFunction.fst (hunk 8 above)
[the remark this answers is not in the document]
Comment (cwaldm) [reply src/lib/utils/DY.Lib.SplitFunction.fst]: Very nice!
I think there are more `l`s and `lpreds` in State.Tagged and Event.Typed that could be renamed to `tagged_local_preds` as well. (not urgent though)

**Comment by cwaldm** [comment]
It looks like a nice generalization!
Does it make sense to provide a "simplified" version for cases where we actually don't have sets (but only singletons), like for State.Tagged and Event.Typed?
I'm thinking of hiding the common part:
```
tag_set_t = string;
tag_t = string;
is_disjoint = unequal;
tag_belong_to = (fun dtag tag -> dtag = tag);
cant_belong_to_disjoint_sets = (fun dtag tag1 tag2 -> ());
```
(where of course we should replace `string` by any `a`)

File: src/lib/utils/DY.Lib.SplitFunction.fst (hunk 1 above)
Comment (cwaldm) [reply src/lib/utils/DY.Lib.SplitFunction.fst]: I agree, that having `for_allP` (and other basic lemmas) in the F* library would be the best option, but also that it's not the most convenient.
For now, I'd prefer to have a `DY.Utils.List` module in DY* where we can collect basic functions (for example `for_allP`, `memP_map`, ...). I sometimes have to define functions/lemmas on Lists and it would be nice to have a place where we collect all of those.
But we can do this later in a separate step.

File: src/lib/utils/DY.Lib.SplitFunction.fst (hunk 2 above)
Comment (cwaldm) [reply src/lib/utils/DY.Lib.SplitFunction.fst]: Thanks!

File: src/lib/state/DY.Lib.State.Tagged.fst (hunk 5 above)
Comment (cwaldm) [reply src/lib/state/DY.Lib.State.Tagged.fst:111]: If we increase ifuel to 2, we can get rid of the whole match. Does anything speak against doing that?

File: src/lib/utils/DY.Lib.SplitFunction.fst (hunk 10 above)
Comment (TWal) [reply src/lib/utils/DY.Lib.SplitFunction.fst]: I updated the comment. We can show equivalence but we don't need the other direction.

File: src/lib/utils/DY.Lib.SplitFunction.fst (hunk 11 above)
Comment (TWal) [reply src/lib/utils/DY.Lib.SplitFunction.fst]: Done

File: src/lib/state/DY.Lib.State.Tagged.fst (hunk 5 above)
Comment (TWal) [reply src/lib/state/DY.Lib.State.Tagged.fst]: I did that, in general I prefer keeping fuel / ifuel as low as possible, but I guess it's okay here because the proof is easy enough for the SMT

**Comment by TWal** [comment]
It's a good idea, I implemented the "hiding" tag and tag sets part

**Review by cwaldm — APPROVED** [review]
Very nice work!

**Comment by TWal** [comment]
Thanks!
